File: coral-go/internal/background/idle_detector.go
package background

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"time"

	"github.com/cdknorow/coral/internal/store"
)

const needsInputThreshold = 300 // 5 minutes

// IdleDetector detects agents waiting for input and creates webhook deliveries.
type IdleDetector struct {
	taskStore    *store.TaskStore
	webhookStore *store.WebhookStore
	interval     time.Duration
	logger       *slog.Logger
	discoverFn   func(ctx context.Context) ([]AgentInfo, error)
	notified     map[string]bool // Track which agents we've already notified
}

// NewIdleDetector creates a new IdleDetector.
func NewIdleDetector(taskStore *store.TaskStore, webhookStore *store.WebhookStore, interval time.Duration) *IdleDetector {
	return &IdleDetector{
		taskStore:    taskStore,
		webhookStore: webhookStore,
		interval:     interval,
		logger:       slog.Default().With("service", "idle_detector"),
		notified:     make(map[string]bool),
	}
}

// SetDiscoverFn sets a custom agent discovery function.
func (d *IdleDetector) SetDiscoverFn(fn func(ctx context.Context) ([]AgentInfo, error)) {
	d.discoverFn = fn
}

// Run starts the detection loop.
func (d *IdleDetector) Run(ctx context.Context) error {
	ticker := time.NewTicker(d.interval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
			if err := d.RunOnce(ctx); err != nil {
				d.logger.Error("detection error", "error", err)
			}
		}
	}
}

// RunOnce performs a single idle detection pass.
func (d *IdleDetector) RunOnce(ctx context.Context) error {
	configs, err := d.webhookStore.ListWebhookConfigs(ctx, true)
	if err != nil || len(configs) == 0 {
		return err
	}

	agents, err := d.discoverFn(ctx)
	if err != nil {
		return err
	}

	sessionIDs := make([]string, 0, len(agents))
	for _, a := range agents {
		if a.SessionID != "" {
			sessionIDs = append(sessionIDs, a.SessionID)
		}
	}
	latestEvents, err := d.taskStore.GetLatestEventTypes(ctx, sessionIDs)
	if err != nil {
		return err
	}

	for _, agent := range agents {
		evPair, ok := latestEvents[agent.SessionID]
		latestEv := ""
		if ok {
			latestEv = evPair[0]
		}
		waiting := latestEv == "stop" || latestEv == "notification"

		if !waiting {
			delete(d.notified, agent.AgentName)
			continue
		}

		// Check log file staleness
		logPath := fmt.Sprintf("%s/%s_coral_%s.log", os.TempDir(), agent.AgentType, agent.SessionID)
		info, err := os.Stat(logPath)
		if err != nil {
			continue
		}
		staleness := time.Since(info.ModTime()).Seconds()
		if staleness < needsInputThreshold {
			continue
		}

		if d.notified[agent.AgentName] {
			continue
		}

		for _, cfg := range configs {
			if cfg.AgentFilter != nil && *cfg.AgentFilter != "" && *cfg.AgentFilter != agent.AgentName {
				continue
			}
			minutes := int(staleness / 60)
			d.webhookStore.CreateWebhookDelivery(ctx,
				cfg.ID, agent.AgentName, "needs_input",
				fmt.Sprintf("Agent needs input — waiting for %d minutes", minutes),
				&agent.SessionID)
		}
		d.notified[agent.AgentName] = true
	}

	return nil
}

File: coral-python/coral/background_tasks/idle_detector.py
"""Needs-input detector — sends a single webhook when an agent waits for input too long."""

from __future__ import annotations

import asyncio
import logging
import time
from pathlib import Path

from coral.tools.session_manager import discover_coral_agents

log = logging.getLogger(__name__)

NEEDS_INPUT_THRESHOLD = 300  # 5 minutes


class IdleDetector:
    """Periodically checks for agents waiting for input and creates webhook deliveries."""

    def __init__(self, store) -> None:
        self._store = store
        # Track which agents we've already notified so we only fire once per
        # waiting period. Cleared when the agent leaves the "needs input" state.
        self._notified: set[str] = set()

    async def run_forever(self, interval: float = 60) -> None:
        while True:
            try:
                await self.run_once()
            except Exception:
                log.exception("IdleDetector error")
            await asyncio.sleep(interval)

    async def run_once(self) -> dict[str, int]:
        """Check all agents. Returns {"notifications": n}."""
        configs = await self._store.list_webhook_configs(enabled_only=True)
        if not configs:
            return {"notifications": 0}

        agents = await discover_coral_agents()
        session_ids = [a["session_id"] for a in agents if a.get("session_id")]
        latest_events = await self._store.get_latest_event_types(session_ids)

        # Skip sleeping sessions — they're intentionally idle
        try:
            all_live = await self._store.get_all_live_sessions()
            sleeping_sids = {s["session_id"] for s in all_live if s.get("is_sleeping")}
        except Exception:
            sleeping_sids = set()

        count = 0
        active_waiting: set[str] = set()

        for agent in agents:
            name = agent["agent_name"]
            sid = agent.get("session_id")
            if sid and sid in sleeping_sids:
                self._notified.discard(name)
                continue
            ev_tuple = latest_events.get(sid) if sid else None
            latest_ev = ev_tuple[0] if ev_tuple else None
            waiting = latest_ev == "notification"

            if not waiting:
                # Agent is active — clear notification state
                self._notified.discard(name)
                continue

            # Agent is waiting for input — check staleness via stat (no file read)
            try:
                staleness = time.time() - Path(agent["log_path"]).stat().st_mtime
            except OSError:
                staleness = 0

            if staleness < NEEDS_INPUT_THRESHOLD:
                continue

            active_waiting.add(name)

            if name in self._notified:
                continue  # Already sent notification for this waiting period

            # Send one notification per enabled webhook
            for cfg in configs:
                if cfg["agent_filter"] and cfg["agent_filter"] != name:
                    continue
                minutes = int(staleness // 60)
                await self._store.create_webhook_delivery(
                    webhook_id=cfg["id"],
                    agent_name=name,
                    session_id=sid,
                    event_type="needs_input",
                    event_summary=f"Agent needs input — waiting for {minutes} minutes",
                )
                count += 1

            self._notified.add(name)

        return {"notifications": count}
